### ISSUE TITLE
DP sensitivity calculation

### ISSUE THREAD

**Issue by eardic** [wenzhu23333/Differential-Privacy-Based-Federated-Learning#14]
Hi,

I don't understand how you calculate the sensitivity of DP to determine the noise level for the gradients. For example for a local dp process, can you recommend a way to compute sensitivity for this code block?

class Laplace:

    def __init__(self, *, epsilon, delta=0.0, sensitivity):
        self.scale = float(sensitivity) / (float(epsilon) - np.log(1 - float(delta)))

    def compute_noise(self, size):
        return torch.tensor(np.random.laplace(loc=0.0, scale=self.scale, size=size))

In literature, it is said that sensitivity is computed by measuring the change in the output of the model by removing or changing an input.

**Comment by wenzhu23333** [comment]
It seems that you are using Laplace mechanism, 
here, delta = 0, and the sensitivty can be compute:
if you add noise to parameter: 
`def cal_sensitivity(lr, clip, dataset_size):
    return 2 * lr * clip / dataset_size`
if add noise to gradient:
`def cal_sensitivity(lr, clip, dataset_size):
    return 2 * clip / dataset_size`
The reason why sensitivity can be compute in this way can refer to these papers:
[1] Y. Zhou et al., "Optimizing the Numbers of Queries and Replies in Convex Federated Learning with Differential Privacy," in IEEE Transactions on Dependable and Secure Computing, doi: 10.1109/TDSC.2023.3234599.
[2] Y. Zhou, et al.,"Exploring the Practicality of Differentially Private Federated Learning: A Local Iteration Tuning Approach" in IEEE Transactions on Dependable and Secure Computing, doi: 10.1109/TDSC.2023.3325889

**Comment by eardic** [comment]
I use FedML framework to train FL models for MNIST and CIFAR10 datasets. Before sending model weights to server, the aim is to add laplace noise to provide DP, but i am not sure if this '2 * clip / dataset_size ' assumption also holds for non-iid dataset. In addition, what should be the value of clip in this formula ? 

If epsilon=1, delta=0, clip=1 and local dataset_size=30 then sensitivity is 1/15=0.06 which generates too much noise for weights and model does not converge during training. What should i do in this case?

Thanks in advance.

**Comment by wenzhu23333** [comment]
In federated learning, the convergence of the Laplace mechanism is often challenging, primarily because the L1 norm tends to be very large, and the clipping process can significantly reduce accuracy. Many papers opt for the Gaussian Mechanism in the context of federated learning. While some studies utilize the Laplace mechanism, they usually do so with very small datasets or models, as demonstrated in papers like [this one](https://arxiv.org/pdf/2101.04163.pdf).

Therefore, I believe it represents an open problem. The loss of convergence could be attributed to two main factors:

An overly restrictive privacy budget;
Excessive information loss due to gradient clipping.
In summary, the only way to enhance accuracy is to strike a balance between the privacy budget and the clipping threshold. This means, firstly, identifying the optimal clipping bound and then adjusting the privacy budget to a reasonable level.

**Comment by eardic** [comment]
Hi again,

Do you find DP-SGD, which applies DP to gradients, more robust and convergible instead of applying DP to model weights? I use multi task auto encoder architecture but having convergence problems when using DP for model weights.  

Do you know how to calculate the sensitivity of gaussian for DP-SGD?

**Comment by wenzhu23333** [comment]
Paper [1] details the application of DP-SGD (Differentially Private Stochastic Gradient Descent) noise to the gradient in Federated Learning (FL), providing insights on how to calculate the sensitivity for DP-SGD.

I believe the crux of the matter isn't about adding noise to the gradient or the model weights per se, as these two methods are fundamentally equivalent. When noise is added to model weights, it's crucial to compute the sensitivity of the "model weight". This sensitivity calculation reveals that it includes the learning rate, which is typically a small value. Conversely, if noise is added to the gradient, the calculation pertains to the sensitivity of the "gradient", which notably does not incorporate the learning rate.

In essence, using the same sensitivity measure for adding noise to both weights and gradients may not be appropriate. Such an approach could lead to convergence issues when adding noise to weights, as they should have a significantly smaller sensitivity.

Reference:
[1] Y. Zhou et al., "Optimizing the Numbers of Queries and Replies in Convex Federated Learning with Differential Privacy," in IEEE Transactions on Dependable and Secure Computing, doi: 10.1109/TDSC.2023.3234599.

**Comment by eardic** [comment]
Hi,

As I understand from [this one](https://arxiv.org/pdf/2101.04163.pdf), the gaussian-based DP noise for a client must be as follows:

Let `W_i` be the model weights of a client i.
`sensitivity = learning_rate * local_training_epochs * gradient_clip`
`tl = total_comm_round * (client_num_per_round / client_num_in_total)`
`std = (sensitivity ** 2) * (tl * np.log(1 / delta)) / (epsilon ** 2)`
`M=torch.normal(mean=0, std=std, size=W_i.shape)`
`W_i = W_i + M `

Is this right?

**Comment by wenzhu23333** [comment]
This is an unpublished manuscript, the public version of this manuscript is:
Y. Zhou, et al.,"Exploring the Practicality of Differentially Private Federated Learning: A Local Iteration Tuning Approach" in IEEE Transactions on Dependable and Secure Computing, doi: 10.1109/TDSC.2023.3325889.

**Comment by eardic** [comment]
This only contains Laplacian noise. Where can I find the Gaussian version of this?

Thanks.

**Comment by eardic** [comment]
Hi wenzhu,

Is there an up limit of epsilon value for laplacian e-dp ? Due to convergence issues, i need to set values such 10^5 or 10^6 for epsilon.

Thanks.

**Comment by wenzhu23333** [comment]
> Hi wenzhu,
> 
> Is there an up limit of epsilon value for laplacian e-dp ? Due to convergence issues, i need to set values such 10^5 or 10^6 for epsilon.
> 
> Thanks.

This value seems excessively high. While theoretically, the epsilon in differential privacy could be any number due to the application of composition rules, in papers I've encountered, it is usually kept under 100 and rarely exceeds 10^3. An epsilon value above 100 is still considered quite large. Typically, epsilon should be between 1 and 10, and it's advisable to keep this value as low as possible for enhanced privacy protection.
